compiler/noirc_frontend/src/ownership/mod.rs

```diff
@@ -260,180 +344,321 @@ fn recur_on_parameter<'typ>(
-}
+    fn handle_unary(&mut self, expr: &mut Expression) {
+        let unary = match expr {
+            Expression::Unary(unary) => unary,
+            other => panic!("handle_unary given non-unary expression: {other}"),
+        };
 
-fn handle_while(while_expr: &mut crate::monomorphization::ast::While) {
-    handle_expression(&mut while_expr.condition);
-    handle_expression(&mut while_expr.body);
-}
+        // Don't clone `rhs` if this is a reference or dereference exprssion.
+        // - If this is a reference expression `&rhs`, `rhs` by definition shouldn't be cloned
+        // - If this is `*rhs` we're going to clone the extracted element instead
+        if self.experimental_ownership_feature
+            && matches!(unary.operator, UnaryOp::Reference { .. } | UnaryOp::Dereference { .. })
+        {
+            self.handle_reference_expression(&mut unary.rhs);
+        } else {
+            self.handle_expression(&mut unary.rhs);
+        }
 
-fn handle_if(if_expr: &mut crate::monomorphization::ast::If) {
-    handle_expression(&mut if_expr.condition);
-    handle_expression(&mut if_expr.consequence);
-    if let Some(alt) = &mut if_expr.alternative {
-        handle_expression(alt);
+        if self.experimental_ownership_feature
+            && matches!(unary.operator, UnaryOp::Dereference { .. })
+            && contains_array_or_str_type(&unary.result_type)
+        {
+            clone_expr(expr);
+        }
     }
-}
 
-fn handle_match(match_expr: &mut crate::monomorphization::ast::Match) {
-    for case in &mut match_expr.cases {
-        handle_expression(&mut case.branch);
+    fn handle_binary(&mut self, binary: &mut crate::monomorphization::ast::Binary) {
+        self.handle_expression(&mut binary.lhs);
+        self.handle_expression(&mut binary.rhs);
     }
 
-    if let Some(default_case) = &mut match_expr.default_case {
-        handle_expression(default_case);
+    /// - Extracting an array from another array (`let inner: [_; _] = array[0];`):
+    ///   - Extracting a nested array from its outer array will always increment the reference count
+    ///     of the nested array.
+    fn handle_index(&mut self, index_expr: &mut Expression) {
+        let crate::monomorphization::ast::Expression::Index(index) = index_expr else {
+            panic!("handle_index should only be called with Index nodes");
+        };
+
+        let mut clone_result = !self.experimental_ownership_feature;
+
+        if self.experimental_ownership_feature {
+            // Don't clone the collection, cloning only the resulting element is cheaper
+            self.handle_reference_expression(&mut index.collection);
+            clone_result = true;
+        } else {
+            self.handle_expression(&mut index.collection);
+        }
+
+        self.handle_expression(&mut index.index);
+
+        if clone_result && contains_array_or_str_type(&index.element_type) {
+            clone_expr(index_expr);
+        }
```

Comment:
```suggestion
        if self.experimental_ownership_feature {
            // Don't clone the collection, cloning only the resulting element is cheaper
            self.handle_reference_expression(&mut index.collection);
        } else {
            self.handle_expression(&mut index.collection);
        }

        self.handle_expression(&mut index.index);

        if contains_array_or_str_type(&index.element_type) {
            clone_expr(index_expr);
        }
```
clone_result will always be true if experimental_ownership_feature is false. Then if experimental_ownership_feature is true we end up setting clone_result to true anyway. Looks like we can get rid of this `clone_result` flag.